### PR TITLE
Debug the failed transaction in SUT plugin

### PR DESCRIPTION
**Pull request by Aniket-Engg**

### CODE REVIEW

File: libs/remix-ui/debugger-ui/src/lib/debugger-ui.tsx [hunk 1]
```diff
@@ -181,7 +181,7 @@ export const DebuggerUI = (props: DebuggerUIProps) => {
       return
     }
 
-    const web3 = state.opt.debugWithLocalNode ? await debuggerModule.web3() : await debuggerModule.getDebugWeb3()
+    const web3 = state.opt.debugWithLocalNode ? await debuggerModule.web3() : (optWeb3 || await debuggerModule.getDebugWeb3())
```
Comment (yann300) [inline libs/remix-ui/debugger-ui/src/lib/debugger-ui.tsx:184]: I think that if `optWeb3` is present, it should have predominance over `await debuggerModule.web3()`

File: libs/remix-tests/src/runTestSources.ts [hunk 2]
```diff
@@ -5,133 +5,142 @@ import { deployAll } from './deployer'
...
 
-const createWeb3Provider = async function () {
-  const web3 = new Web3()
-  const provider: any = new Provider()
-  await provider.init()
-  web3.setProvider(provider)
-  extend(web3)
-  return web3
-}
+export class UnitTestRunner {
```
Comment (yann300) [inline libs/remix-tests/src/runTestSources.ts:16]: any chance we can discard `runTestFile` and use this instead?

File: apps/debugger/src/app/debugger-api.ts [hunk 3]
```diff
@@ -121,9 +121,10 @@ export const DebuggerApiMixin = (Base) => class extends Base {
     return await debug.debugger.traceManager.getTrace(hash)
   }
 
-  debug (hash) {
+  debug (hash, web3?) {
     this.debugHash = hash
-    if (this.onDebugRequestedListener) this.onDebugRequestedListener(hash)
+    if(web3) remixDebug.init.extendWeb3(web3)
```
Comment (LianaHus) [inline apps/debugger/src/app/debugger-api.ts:126]: space after if

File: apps/remix-ide/src/app/tabs/test-tab.js [hunk 4]
```diff
@@ -242,13 +264,30 @@ module.exports = class TestTab extends ViewPlugin {
     } else if (result.type === 'testFailure') {
       if (result.hhLogs && result.hhLogs.length) this.printHHLogs(result.hhLogs, result.value)
       if (!result.assertMethod) {
+        let debugBtn = yo``
+        if (result.errMsg.includes('Transaction has been reverted by the EVM')) {
```
Comment (yann300) [inline apps/remix-ide/src/app/tabs/test-tab.js:268]: I think the option to debug should be displayed in all cases (e.g if tests fails or not)

File: apps/remix-ide/src/app/tabs/test-tab.js [hunk 5]
```diff
@@ -242,13 +264,30 @@ module.exports = class TestTab extends ViewPlugin {
...
+          const txHash = JSON.parse(result.errMsg.replace('Transaction has been reverted by the EVM:', '')).transactionHash
+          const { web3 } = result
+          debugBtn = yo`<div
```
Comment (yann300) [inline apps/remix-ide/src/app/tabs/test-tab.js:271]: perhaps the debug button should have the same format as of the other actions in the tab (e.g an icon followed by a label).

File: libs/remix-tests/src/runTestSources.ts (hunk 2 above)
Comment (Aniket-Engg) [reply libs/remix-tests/src/runTestSources.ts:16]: That is there for CLI usage. I would like to do that in a separate PR. Will be tracked by https://github.com/ethereum/remix-project/issues/1561

File: apps/remix-ide/src/app/tabs/test-tab.js (hunk 5 above)
Comment (Aniket-Engg) [reply apps/remix-ide/src/app/tabs/test-tab.js:271]: For now, this looks very composed there on the card. @LianaHus also agreed on it as it is now.

File: apps/debugger/src/app/debugger-api.ts (hunk 3 above)
Comment (Aniket-Engg) [reply apps/debugger/src/app/debugger-api.ts]: updated

File: libs/remix-ui/debugger-ui/src/lib/debugger-ui.tsx (hunk 1 above)
Comment (Aniket-Engg) [reply libs/remix-ui/debugger-ui/src/lib/debugger-ui.tsx]: updated

File: apps/remix-ide/src/app/tabs/test-tab.js (hunk 4 above)
Comment (yann300) [reply apps/remix-ide/src/app/tabs/test-tab.js:268]: after discussion, this will be handled in an another PR.

Comment (Aniket-Engg) [reply apps/remix-ide/src/app/tabs/test-tab.js:268]: Issue to track: https://github.com/ethereum/remix-project/issues/1562

File: libs/remix-tests/src/runTestSources.ts [hunk 6]
```diff
@@ -5,133 +5,142 @@ import { deployAll } from './deployer'
 import { runTest } from './testRunner'
 
 import Web3 from 'web3'
+import EventManager from './lib/eventManager'
```
Comment (yann300) [inline libs/remix-tests/src/runTestSources.ts:8]: Would it be possible to use `const EventEmitter = require('events')` instead?

Comment (Aniket-Engg) [reply libs/remix-tests/src/runTestSources.ts:8]: will check

Comment (Aniket-Engg) [reply libs/remix-tests/src/runTestSources.ts]: done

**Review by yann300 — APPROVED** [review]

**Pull request merged**